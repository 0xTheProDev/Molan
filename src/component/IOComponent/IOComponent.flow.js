// @flow
export type IOComponentProps = {
  status: number,
  input?: string,
  output?: string,
};

export type IOComponentInternalProps = {
  color: string,
  status: string,
  input?: string,
  output?: string,
  StatusText: string,
  CustomInputText: string,
  CustomOutputText: string,
  InputFieldPlaceholder: string,
  OutputFieldPlaceholder: string,
};
