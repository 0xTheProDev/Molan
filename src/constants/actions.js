export const ACTIONS = {

  /** To Get the Health Status of Runner Application */
  GET_STATUS: 'GET_STATUS',

  /** To Submit Code and Language Action */
  SUBMIT_CODE: 'SUBMIT_CODE',
};

export default ACTIONS;
