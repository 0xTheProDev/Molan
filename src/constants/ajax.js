/** Base URL to Runner Application (e.g., `localhost:5001`) */
export const BASE_URL = 'http://localhost:5001/molan-ide/us-central1/default-app';

/** HTTP Methods */
export const METHODS = Object.freeze({
  GET:      'GET',
  POST:     'POST',
  PUT:      'PUT',
  DELETE:   'DELETE',
});

/** Allowed HTTP Methods for Ajax */
export const ALLOWED_METHODS = Object.freeze([
  METHODS.GET,
  METHODS.POST,
]);

/** Error string for Invalid Method */
export const METHOD_NOT_ALLOWED = 'Method not allowed for Ajax Request';
