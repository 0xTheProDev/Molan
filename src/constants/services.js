// @flow
export const SERVICE_STATUS = Object.freeze({
  LOADING: 0,
  READY: 1,
  ERROR: 2,
});

export const COMPILER_STATUS = Object.freeze({
  INPROGRESS: 0,
  COMPILATION_ERROR: 1,
  RUNTIME_ERROR: 2,
  SUCCESS: 3,
});
