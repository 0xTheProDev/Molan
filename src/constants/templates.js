// @flow
import { LanguageTemplate, LanguageTemplateOptions } from 'types/template.flow';

const c_template =
`#include <stdio.h>

int main(void) {
  /* Your Code Here */
  return 0;
}
`;

const cpp_template =
`#include <iostream>
using namespace std;

int main() {
  // Your Code Here
  return 0;
}
`;

const java_template =
`import java.io.*;

class Molan {
  public static void main (String[] args) {
    // Your Codes Here
  }
}
`;

const py_template =
`# Write Codes Here
`;

const mips_template =
`    .text
main:
  # Your Codes Here
`;

const ver_template =
`/** Workbench Module */
module wb;
  /** Main Simulation Here */
endmodule
`;

export const LANGUAGES = Object.freeze([
  'c',
  'cpp',
  'java',
  'python',
  'python3',
  'mips',
  'verilog',
]);

export const COMPILER_OPTIONS: LanguageTemplateOptions = Object.freeze({
  c: 'C (GCC 6.3)',
  cpp: 'C++ (G++ 6.3)',
  java: 'Java (OpenJDK 0.8)',
  python: 'Python (Python 2.7)',
  python3: 'Python (Python 3.6)',
  mips: 'MIPS-IV (Simulator 0.4)',
  verilog: 'Verilog (iVerilog 10.0)',
});

export const LANGUAGE_TEMPLATES: LanguageTemplateOptions = Object.freeze({
  c: c_template,
  cpp: cpp_template,
  java: java_template,
  python: py_template,
  python3: py_template,
  mips: mips_template,
  verilog: ver_template,
});

export const MIME_TYPES: LanguageTemplateOptions = Object.freeze({
  c: 'text/x-csrc',
  cpp: 'text/x-c++src',
  java: 'text/x-java',
  python: 'text/x-python',
  python3: 'text/x-python',
  mips: 'text/plain',
  verilog: 'text/plain',
  text: 'text/plain',
});

export const FILE_EXTENSIONS: LanguageTemplateOptions = Object.freeze({
  c: '.c',
  cpp: '.cxx',
  java: '.java',
  python: '.py',
  python3: '.py',
  mips: '.s',
  verilog: '.v',
  text: '.txt',
});

const LANGUAGE_OPTIONS: LanguageTemplate[] = LANGUAGES.reduce((acc: LanguageTemplate[], currentLanguage: string) => {
  acc.push({
    language: currentLanguage,
    name: COMPILER_OPTIONS[currentLanguage],
    code: LANGUAGE_TEMPLATES[currentLanguage],
    mime: MIME_TYPES[currentLanguage],
    extenstion: FILE_EXTENSIONS[currentLanguage],
  });
  return acc;
}, []);

export default Object.freeze(LANGUAGE_OPTIONS);
