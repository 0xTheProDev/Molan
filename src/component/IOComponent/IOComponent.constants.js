// @flow
import { COMPILER_STATUS } from 'constants/services';

export const SUCCESS_STATUS = 'Success';
export const SUCCESS_STATUS_COLOR = 'green';

export const RUNTIME_ERROR_STATUS = 'Runtime Error';
export const RUNTIME_ERROR_STATUS_COLOR = 'red';

export const COMPILE_ERROR_STATUS = 'Compilation Error';
export const COMPILE_ERROR_STATUS_COLOR = 'yellow';

export const INPROGRESS_STATUS = 'Processing';
export const INPROGRESS_STATUS_COLOR = 'teal';

export const STATUS_TEXT_MAPPING = Object.freeze({
  [COMPILER_STATUS.INPROGRESS]: INPROGRESS_STATUS,
  [COMPILER_STATUS.COMPILATION_ERROR]: COMPILE_ERROR_STATUS,
  [COMPILER_STATUS.RUNTIME_ERROR]: RUNTIME_ERROR_STATUS,
  [COMPILER_STATUS.SUCCESS]: SUCCESS_STATUS,
});

export const STATUS_COLOR_MAPPING = Object.freeze({
  [COMPILER_STATUS.INPROGRESS]: INPROGRESS_STATUS_COLOR,
  [COMPILER_STATUS.COMPILATION_ERROR]: COMPILE_ERROR_STATUS_COLOR,
  [COMPILER_STATUS.RUNTIME_ERROR]: RUNTIME_ERROR_STATUS_COLOR,
  [COMPILER_STATUS.SUCCESS]: SUCCESS_STATUS_COLOR,
});
