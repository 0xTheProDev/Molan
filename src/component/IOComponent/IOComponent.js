// @flow
import React from 'react';
import { Grid, Form, TextArea } from 'semantic-ui-react';

import { IOComponentInternalProps } from './IOComponent.flow';

export function IOComponent(props: IOComponentInternalProps) {
  const {
    color,
    status,
    input,
    output,
    StatusText,
    CustomInputText,
    CustomOutputText,
    InputFieldPlaceholder,
    OutputFieldPlaceholder,
  } = props;

  return (
    <Grid>
      <Grid.Row>
        <Grid.Column color={color}>
          <label htmlFor="compiler_status">{StatusText}</label>
          <span id="compiler_status">{status}</span>
        </Grid.Column>
      </Grid.Row>
      <Grid.Row columns={1}>
        <Grid.Column stretched>
          <Form>
            <label htmlFor="custom_input">{CustomInputText}</label>
            <TextArea
              disabled
              autoHeight
              value={input}
              placeholder={InputFieldPlaceholder}
              id="custom_input"
            />
          </Form>
        </Grid.Column>
      </Grid.Row>
      <Grid.Row columns={1}>
        <Grid.Column stretched>
          <Form>
            <label htmlFor="custom_output">{CustomOutputText}</label>
            <TextArea
              disabled
              autoHeight
              value={output}
              placeholder={OutputFieldPlaceholder}
              id="custom_output"
            />
          </Form>
        </Grid.Column>
      </Grid.Row>
    </Grid>
  );
}

IOComponent.defaultProps = {
  StatusText: 'Status',
  CustomInputText: 'Custom Input',
  CustomOutputText: 'Custom Output',
  InputFieldPlaceholder: 'No Input Provided',
  OutputFieldPlaceholder: 'No Output to Show',
};
