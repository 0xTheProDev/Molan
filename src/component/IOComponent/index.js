// @flow
import React from 'react';

import { IOComponentProps } from './IOComponent.flow';
import { IOComponent } from './IOComponent';
import { STATUS_TEXT_MAPPING, STATUS_COLOR_MAPPING } from './IOComponent.constants';

export default function IOComponentFactory({ status, input, output }: IOComponentProps) {
  const Color = STATUS_COLOR_MAPPING[status];
  const Status = STATUS_TEXT_MAPPING[status];

  return (
    <IOComponent
      color={Color}
      status={Status}
      input={input}
      output={output}
    />
  );
}

export { IOComponentProps, IOComponentFactory as IOComponent };
