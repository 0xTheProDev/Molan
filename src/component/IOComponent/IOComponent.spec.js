import React from 'react';

import { IOComponent } from './IOComponent';
import { COMPILE_ERROR_STATUS, COMPILE_ERROR_STATUS_COLOR } from './IOComponent.constants';

describe('Component::IOComponent::IOComponent', () => {
  let minProps;

  beforeAll(() => {
    minProps = {
      color: COMPILE_ERROR_STATUS_COLOR,
      status: COMPILE_ERROR_STATUS,
      input: 'Some Input Here',
      output: 'Some Output Here',
    };
  });

  it('should render with base structure', () => {
    const wrapper = shallow(<IOComponent {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
