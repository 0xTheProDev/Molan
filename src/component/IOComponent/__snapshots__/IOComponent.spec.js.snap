// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component::IOComponent::IOComponent should render with base structure 1`] = `
<Grid>
  <GridRow>
    <GridColumn
      color="yellow"
    >
      <label
        htmlFor="compiler_status"
      >
        Status
      </label>
      <span
        id="compiler_status"
      >
        Compilation Error
      </span>
    </GridColumn>
  </GridRow>
  <GridRow
    columns={1}
  >
    <GridColumn
      stretched={true}
    >
      <Form
        as="form"
      >
        <label
          htmlFor="custom_input"
        >
          Custom Input
        </label>
        <TextArea
          as="textarea"
          autoHeight={true}
          disabled={true}
          id="custom_input"
          placeholder="No Input Provided"
          rows={3}
          value="Some Input Here"
        />
      </Form>
    </GridColumn>
  </GridRow>
  <GridRow
    columns={1}
  >
    <GridColumn
      stretched={true}
    >
      <Form
        as="form"
      >
        <label
          htmlFor="custom_output"
        >
          Custom Output
        </label>
        <TextArea
          as="textarea"
          autoHeight={true}
          disabled={true}
          id="custom_output"
          placeholder="No Output to Show"
          rows={3}
          value="Some Output Here"
        />
      </Form>
    </GridColumn>
  </GridRow>
</Grid>
`;
